htcanalyze/htcanalyze.py

```diff
@@ -64,28 +64,44 @@ def manage_thresholds(self, resources: dict) -> dict:
 
         The important part is that the keywords exist
         "Usage", "Requested"
-        marks everything with the colors
-        (green, yellow, red) - (good, warning, bad)
 
-        :param resources: resource dictionary
+        :param resources: a dict like:
+        {
+            "Resources": ["Cpu", "Disk", "Memory"]
+            "Usage": [0, 13, 144]
+            "Requested": [1, 1000, 6000]
+            "Allocated": [1, 3000, 134900]
+        }
         :return: resources with colors on usage column
         """
         # change to list, to avoid numpy type errors
         resources.update(Usage=list(resources["Usage"]))
-        # TODO: What does i do? Making resources['Requested'] a dict would be more descriptive
-        for i, _ in enumerate(resources['Resources']):
+        n_resources = len(resources['Resources'])
+
```

Comment:
Style nit: when defining a variable for a loop, traditionally there is no empty line separating the assignment and the loop.

I don't see `n_resources` used again later, so this could be collapsed to:

```
for i in range(len(resources['Resources'])):
```

Possible future work could explore whether a list-of-dicts is a better way to store this information rather than a dict-of-lists. That way each resource would be self contained (e.g. `[ {'Resource': 'Cpu', 'Usage': 0, 'Requested': 1, 'Allocated': 1 } ]`), and could use an even more Pythonic (and IMO, mentally easier to parse):

```
for i in resources:
```

Reply:
Initially I used this kind of format (and did not change it yet) because of the earlier use of the tool: tabulate. Just so you know, that I am not an alien using the most complicated structures, you can think of 🌝.  
I am now ambivalent about the idea of a Ressource class mentioned by @nhjjreuter above, or this one.
I think to stick with a Ressource class would simplify this even more, but might increase the speed of this facility.
What do you think ?

Reply:
Style nit on Alex' suggestion: `for i in resources` should obviously be `for resource in resources` ;)

The speed is not relevant here. It is a matter of microseconds. You can keep the tabular structure, and I could even help you write a function that converts a list of dicts to a dict of lists, so you get the tabular structure back. With the approach I wrote, you can approach the tables line-by-line as well.

The idea is to create an instance of the `Resource` object that contains all the data for a resource. If you need to modify it (i.e., add a warning level for the resource, compute some derived value, etc.) you can do so inside of the `Resource` instance. Any modifications will then be grouped together, rather than spread out over the script.

In your current script you change the value of `resources['Usage']` to a color-coded string. Now you've prevent yourself from working with the usage value, if the need ever arises. You've also changed the type of the variable from an integer (or float) to a string. Using the `Resource` class approach, you instead format the textual output later in the code, using the `Resource.usage_level` to determine the color tags.

We can discuss the relevance of this in the group meeting if you'd like.

Reply:
Mh okay, I really like this point of view, I will create an Issue on this, because I think this should be done in an other PR assigning the Issue

Reply:
Create Issue for this idea of a Resource class #118

Reply:
And no there is no need to keep this structure, because I'm not using tabulate anymore. I was just to lazy to change it all, so I kept this structure.